[docs] Add hooks API pages for MUI Base

mnajdova opened the pull request:
This PR adds logic for generating APIs for the hooks (currently used in MUI Base only). It generates the input params and return value considering the exported types from the project by using the following convention:

- `${hookName}Parameters` - for the params
- `${hookname}ReturnValue` - for the return value

It also adds annotation above the hooks about where the demos and API pages are (similar to the components).

Example: https://deploy-preview-35828--material-ui.netlify.app/base/api/use-button/
The API pages for the hooks are linked in the API section of the unstyled components: https://deploy-preview-35828--material-ui.netlify.app/base/react-button/#api For this to work, in the headers of the component markdown there is a new `hook` header.

The relevant changes are in the following files:
- [HookApiPage.js](https://github.com/mui/material-ui/pull/35828/files#diff-2a26832c2bc9cc1b9f544a01e8d3d80ecd10a072e97f1ece4cd85505523b6356) - similar to API page, but adjusted for hooks. Here is one example: https://deploy-preview-35828--material-ui.netlify.app/base/api/use-button/
- [HookApiBuilder.ts](https://github.com/mui/material-ui/pull/35828/files#diff-076db56372d049f266a7ca21c4054f4baaa06b92b00f9b03a3694592b145e3c5) - similar to `ComponentApiBuilder`, but adjusted to hooks. The logic for generating the json APIs from interfaces was taken from MUI X, and slightly simplified and adjusted (in MUI Core we don't need to have distinction between different packages pro/premium etc.)
- [buildApi.ts](https://github.com/mui/material-ui/pull/35828/files#diff-d09bb89a0d85b6c953e3e5fc56523e213ac5e1152e2276dc535b61a2020ddf15) - added logic for generating the hooks - invokes the previous mentioned `HookApiBuilder`
- [buildApiUtils.ts](https://github.com/mui/material-ui/pull/35828/files#diff-6a6e2cedd650384468a6c55a3d546c6de8b4a88d07baf5157ce4356c33a75b73) - new utils added relevant for the hooks
- [useButton.type.ts](https://github.com/mui/material-ui/pull/35828/files#diff-5568544e76e2800a393a9758cbb7b95d1d8ed20acdeec53171048db2404e41e1) - for the added return value type

TODOs:
- [x] Add one hook return type & generation
- [x] Add annotation on the hook for API and demos (similar to the components)

After the PR is merged:
- Review all types and add descriptions on all props
- Add ReturnValue interface for all hooks
- Update the useSelect and useListbox types to use generics (similar to useAutocomplete)
- Add documentation for:
  - https://mui.com/material-ui/react-app-bar/#usescrolltrigger-options-trigger
  - https://mui.com/material-ui/react-use-media-query/?

Fixes https://github.com/mui/material-ui/issues/35792

mnajdova commented on docs/src/modules/components/AppNavDrawer.js:
```diff
@@ -217,7 +217,6 @@ function reduceChildRoutes(context) {
   }
 
   const title = pageToTitleI18n(page, t);
```
```suggestion
  const title = pageToTitleI18n(page, t);

```

mnajdova commented on packages/api-docs-builder/ApiBuilders/ComponentApiBuilder.ts:
```diff
@@ -91,7 +91,10 @@ export function writePrettifiedFile(
  */
-async function computeApiDescription(api: ReactApi, options: { host: string }): Promise<string> {
+export async function computeApiDescription(
+  api: { description: ReactApi['description'] },
```
The type is simplified so that it can be used with the hooks generation too. Only the description is used in the function.

mnajdova commented on packages/mui-base/src/AutocompleteUnstyled/useAutocomplete.d.ts:
```diff
@@ -292,6 +292,13 @@ export interface UseAutocompleteProps<
   value?: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>;
 }
 
+export interface UseAutocompleteParameters<
```
Aliasing the interface so that it matches the name pattern used for the hooks parameters.

mnajdova commented on packages/mui-base/src/ButtonUnstyled/useButton.types.ts:
```diff
@@ -39,3 +40,33 @@ export interface UseButtonParameters {
   type?: React.ButtonHTMLAttributes<HTMLButtonElement>['type'];
 }
+
+export interface UseButtonReturnValue {
```
Explicit type added for the `useButton` return value so that we can test out the generation of the API.

mnajdova commented on packages/mui-base/src/TabsListUnstyled/useTabsList.ts:
```diff
@@ -67,8 +67,17 @@ const moveFocus = (
+ *
+ * - [useTabsList API](https://mui.com/base/api/use-tabs-list/)
+ */
+function useTabsList(parameters: UseTabsListParameters) {
```
Converted to function for consistency and simplifying the generation script.

siriwatknp commented:
I'd be nice if the demo links to the hook section.

https://user-images.githubusercontent.com/18292247/213347624-f448dfa8-531f-422b-830a-18bfed2884c6.mov

siriwatknp commented on docs/pages/base/api/use-slot-props.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "inputParams": {},
+  "returnValue": {},
+  "name": "useSlotProps",
+  "filename": "/packages/mui-base/src/utils/useSlotProps.ts",
+  "demos": "<ul></ul>"
+}
```
Should this be excluded from the public? cc @michaldudak

mnajdova replied:
I made the script to not throw if there are no demos, but we will likely need to do another pass and check each of the docs pages.

michaldudak replied:
It could be useful for building custom libraries, so let's leave it in.

mnajdova replied:
Removing it from now, until we have demos for the hook as decided on the backlog grooming. We will need to anyway review the input params descriptions.

mnajdova commented:
> I'd be nice if the demo links to the hook section.

Sure, that's a great point. I will update it. Luckily with all the good work @samuelsycamore did, the pages are consistent and we can do this :)

oliviertassinari commented on docs/pages/base/api/use-button.json:
```diff
@@ -0,0 +1,66 @@
+    "active": {
+      "type": { "name": "boolean", "description": "If `true`, the component is active (pressed)." },
+      "default": "false",
+      "required": true
```
This feels weird in the context of a returned value, e.g.

<img width="151" alt="Screenshot 2023-01-21 at 18 20 18" src="https://user-images.githubusercontent.com/3165635/213879009-0ecb0453-25f7-446b-9e50-327100e21b97.png">

https://deploy-preview-35828--material-ui.netlify.app/base/api/use-button/#returnValue

I think that what MUI X did with the optional values would make more sense in this context: 

<img width="222" alt="Screenshot 2023-01-21 at 18 22 22" src="https://user-images.githubusercontent.com/3165635/213879097-ea99d517-fb58-4d4e-9819-5fb613639142.png">

https://mui.com/x/api/data-grid/grid-filter-model/#properties

mnajdova replied:
Ah nice, yep it looks better, will re-work it 👍

mnajdova replied:
Done with https://github.com/mui/material-ui/pull/35828/commits/140683198cb9a5358e59b255d2a24a103a7fa3fc

oliviertassinari commented on docs/src/modules/components/HookApiPage.js:
```diff
@@ -0,0 +1,286 @@
+const Wrapper = styled('div')({
+  overflow: 'hidden',
+});
+const Table = styled('table')(({ theme }) => {
+  const contentColor =
+    theme.palette.mode === 'dark'
+      ? alpha(theme.palette.primaryDark[900], 1)
+      : 'rgba(255, 255, 255, 1)';
+  const contentColorTransparent =
+    theme.palette.mode === 'dark'
+      ? alpha(theme.palette.primaryDark[900], 0)
+      : 'rgba(255, 255, 255, 0)';
```
This file is duplicated between MUI Core and MUI X. We are adding it a 3rd time. It could be good to normalize this. Maybe it's outside of the scope of this PR, not sure how, but I think we only need it once.

mnajdova replied:
> This file is duplicated between MUI Core and MUI X. We are adding it a 3rd time. It could be good to normalize this. Maybe it's outside of the scope of this PR, not sure how, but I think we only need it once.

Will fix it in this PR.

mnajdova replied:
Done, I've adjusted the API to work with both components & hooks API.

mnajdova replied:
cc @flaviendelangle you could probably use it in the X too.

oliviertassinari replied:
Ah ok, you abstracted the `PropretiesTable`, it helps 👍. I was thinking of going all the way to create a generic API component and have a JavaScript API to control its rendering.

oliviertassinari reviewed (commented):
👍 It's great to have docs for the hooks.

Is there a way it can support https://mui.com/material-ui/react-app-bar/#usescrolltrigger-options-trigger, https://mui.com/material-ui/react-use-media-query/?

mnajdova commented:
> I see that it's broken with the switch too https://deploy-preview-35828--material-ui.netlify.app/base/api/use-switch/#returnValue. Would it make sense to extract these changes out of this PR and to them iteratively?

The return value is not working anywhere. I will be adding them in dedicated PR. I've added it in only one component to test out the script.

mnajdova commented:
> Is there a way it can support https://mui.com/material-ui/react-app-bar/#usescrolltrigger-options-trigger, https://mui.com/material-ui/react-use-media-query/?

Yes, for now I am just adding the hooks docs on MUI Base. I can do it in a follow up PR.

michaldudak commented on packages/api-docs-builder/ApiBuilders/HookApiBuilder.ts:
```diff
@@ -0,0 +1,539 @@
+        return [] as any;
+      }
+
+      // Only keep `default` for bool props if it isn't 'false'.
```
There's no harm for us in being explicit and stating all boolean props' default values, not just non-false ones.

mnajdova replied:
I followed here the logic in the components' props builder. We could fix it separately in both scripts. There was already an issue related to this - https://github.com/mui/material-ui/issues/35822

oliviertassinari commented on docs/src/modules/components/ApiPage.js:
```diff
@@ -1,123 +1,15 @@
-import Alert from '@mui/material/Alert';
 import Typography from '@mui/material/Typography';
 import { useTranslate, useUserLanguage } from 'docs/src/modules/utils/i18n';
+import PropsTable from 'docs/src/modules/components/PropretiesTable';
```
It would be better if it matches

```suggestion
import PropretiesTable from 'docs/src/modules/components/PropretiesTable';
```

mnajdova replied:
Yep, will update 👍

oliviertassinari reviewed (commented):
I can't see anything wrong with the UX 👌, it flows like the component API pages, developers will be happy to have these pages.

mnajdova commented:
> This is MUI X page https://github.com/mui/mui-x/blob/next/docs/src/modules/components/ApiPage.js that is almost exactly the same as https://github.com/mui/material-ui/blob/next/docs/src/modules/components/ApiPage.js.

Ah, I see your point. I haven't looked into this, but I can do a follow-up on this and see what's different between the two. We can have one component that can handle all three scenarios (components, interfaces, hooks), but I woldn't change it in this PR so that we don't have unrelated changes. I will likely need two PRs, one on X side to test out the changes.

michaldudak commented:
The parameters table doesn't format the text as markdown. Specifically, text is backticks isn't formatted as code. See https://deploy-preview-35828--material-ui.netlify.app/base/api/use-button/#parameters and compare to https://deploy-preview-35828--material-ui.netlify.app/base/api/tab-unstyled/#props

Aside from that, it looks good!

michaldudak reviewed (approved):
I couldn't find anything wrong. Good work!

ZeeshanTamboli reviewed (commented):
It looks fantastic! 👌  I did some testing and found a few things. Maybe you are already aware of this. Just listing them so they can be tracked. Some of them can be probably be addressed in this PR.
- There is no link to the demos on the `useAutocomplete` and `useSelect` pages. Its understandable for `useAutocomplete` since there is no `AutocompleteUnstyled`, but `useSelect` can be linked to `SelectUnstyled` page (the hook section)?
- Menu related hooks APIs have `Hooks` section (plural) instead of `Hook` in the demos page.
- Tab/Tabs related hooks don't have their respective hooks section in their unstyled components page.

flaviendelangle commented:
Looks super solid :ok_hand: 

Maybe we could group the entries in the left NavBar to improve the readability.
It's not super urgent since the alphabetical sorting already does most of the work.

mnajdova commented:
> Maybe we could group the entries in the left NavBar to improve the readability.
> It's not super urgent since the alphabetical sorting already does most of the work.

It's an interesting thought, it may depend on whether we want to group them by components vs hooks (maybe add level 2 subgroup), or by concept, for e.g. having ButtonUnstyled & useButton next to each other. Happy to iterate on this later (if we decide to keep the pages vs the tabs).

mnajdova commented:
Nice findings @ZeeshanTamboli 

> There is no link to the demos on the useAutocomplete and useSelect pages. Its understandable for useAutocomplete since there is no AutocompleteUnstyled, but useSelect can be linked to SelectUnstyled page (the hook section)?

Fixed the `useSelect` 👍 

> Menu related hooks APIs have Hooks section (plural) instead of Hook in the demos page.

Added logic to point to #hooks if there are multiple hooks

> Tab/Tabs related hooks don't have their respective hooks section in their unstyled components page.

cc @samuelsycamore we'll need to add it.

ZeeshanTamboli commented on docs/src/modules/components/PropertiesTable.js:
```diff
@@ -0,0 +1,127 @@
+              .replace(/>/g, '&gt')
+              .replace(/"/g, '&quot;')
+              .replace(/'/g, '&#39;');
+            const propDefault = propData.defaul;
```
Oops,
```suggestion
            const propDefault = propData.default;
```
